### ISSUE TITLE
[ISSUE] "We need Ambulance" doesn't get dispatched

### ISSUE THREAD

**Issue by Fogest** [jackbayliss/Mission-Chief-Bot#64]
**Describe the bug**
When there is a "We need: ambulance" it often does not get dispatched. If you take a look at the screenshot this is an example of what built up overnight. None of those calls have an ambulance assigned to them. And looking at the console output I keep seeing two lines in a row saying "Found missing ambulance" yet nothing happens when it finds them. At the time of writing this all 4 of my ambulances are at the station and unused so there are definitely units free.

**Expected behaviour**
Ambulance gets dispatched when one is needed.

**Screenshots**
![image](https://user-images.githubusercontent.com/1163628/95770295-312bbb80-0c87-11eb-936a-bbaa363a78de.png)

** Bot version and region **
Latest bot version on US region.

**Additional context**

**Comment by Fogest** [comment]
Just to add on, some of those calls are requesting things like the 2 firetrucks, and 1 battalion chief as well as an ambulance and none of those resources are sent. However normally units would be sent, however it seems like sometimes it glitches especially when ambulances are involved and doesn't send anyone.

**Comment by jackbayliss** [comment]
Strange for sure- do you have auto share enabled at all in the config? Have you tried re-running the bot ?

**Comment by Fogest** [comment]
Nope, I am in an alliance on my own so I don't share my missions. Re-running doesn't seem to pick them up. It seems to relate to ambulances typically where I see it stop working. If the mission type changes and requires an ambulance they should normally get sent. But if they are on scene and it changes to "We need: Ambulance" then it doesn't send even though the bot seems to pick up that it is missing ambulances.

**Comment by Fogest** [comment]
Just to add on, I also am often seeing "Reported fraud" mission not get any units sent to it. That mission is just 1 x Police Car nothing special happens on it. I have included a screenshot below. I had to manually dispatch all of the ambulances and police in these missions. These all made my mission list "stuck" as I could get no more missions until I manually tended to all these and manually sent ambulances and police.

![image](https://user-images.githubusercontent.com/1163628/96049781-83670b00-0e46-11eb-894a-0829faf52d8d.png)

**Comment by BigBearEvo** [comment]
might want to try adding this into the missing file, it might solve your problem

"ambulance" : [5, 27],

**Comment by Fogest** [comment]
> might want to try adding this into the missing file, it might solve your problem
> 
> "ambulance" : [5, 27],

That seems to already be in there. Also the reported Fraud one works sometimes, just not all the time. The ambulance one is similar in that sometimes it does work and an ambulance gets found being needed and gets sent. 

![image](https://user-images.githubusercontent.com/1163628/96157159-f1134580-0edf-11eb-8b74-0880eb9647d8.png)

**Comment by MC-bot-test** [comment]
Try

"ambulance" : [5, 11, 27],

and see if that fixes it

**Comment by Fogest** [comment]
Hmm, that may fix the industrial fire ones I am not sure. Haven't tried long enough to see yet if any problems happen on the other calls. I know the industrial one is also failing because MissionChief is going by the mission requirements, but I don't have some of the vehicles required and it works without them.

![image](https://user-images.githubusercontent.com/1163628/96197034-e4faa880-0f1e-11eb-904e-94427d12b4d7.png)

For example, I do not have an MCV vehicle even unlocked nor do I have enough firehouses to require it. I also do not yet have a hazmat vehicle. However when I manually dispatch the mission it has no problem completing without hazmat or MVC. So in these cases MC gets stuck because it can't find those units that I have 0 of. However manually I just dispatch the actual requirements.

**Comment by MC-bot-test** [comment]
Agreed,
That is a quirk I have encountered too. You have to amend the individual
mission json files until you get the units, then change them back.

MC will complete a mission without what csn be best described as 'optional'
requirements (as long as you don't have them enabled). The second you
unlock the 'optional' vehicles, the same mission will then not complete
without the optional vehicles.

The bot  doesn't assess what you have - just what is required in the
mission json.

**Comment by Fogest** [comment]
Okay I have updated the mission file for it and will see if that helps then.

**Comment by Fogest** [comment]
It is still getting stuck almost identical to how that first image in the OP looks like even after making the suggested changes and restarting the bot. What is odd is that after making the changes it did start sending people to the Industrial Fire, however when the call initially comes in as another call and then changes to another is when it seems to break. 

In my queue I currently have 2 roof fires that became industrial fires. They both require additional units + ambulances that haven't been sent. There are 2 Tree on the roads that became a Road accidents and now need ambulances but none were sent to either. And then I have a chimney fire -> roof fire -> industrial fire with the same problem as the other industrials. Also got a "reported fraud" in there which still only seems to have a police officer dispatched to it about 25% of the time. Not sure what is going on with the reported fraud one. It is a simple mission that just requires a single police officer right from the beginning and that is all. Sometimes it works, often it doesn't.

**Comment by Fogest** [comment]
Hmm, just to add on, I think I am seeming what part of the problem is. It seems like these roof incidents both required Battalion Chiefs, but then they both required an additional one when they switched to a different call type. Unfortunately other calls already also had battalion chiefs so it left the game in a state where there are no more BC's, but most calls needs one more so it is stuck. I guess I need more BC's. Would be interesting to see the bot detect these cases. Though that is probably a lot more complex. 

So I guess my main issue is related to the odd call where things don't get dispatched like the reported fraud or the odd one where an ambulance is being requested and no ambulance is being sent.

**Comment by jackbayliss** [comment]
Thanks all for letting me know, this is just part of the whole eco-system of the bot. Everything can be changed if required, hence why the mission files are JSON. The idea is that the links.json serves the missions and the missing.json serves the missing vehicles found. Vehicles need to be status 1 or 2 to get despatched, and the whole requirements have to be met in terms of vehicles. If you have a particular mission that's troubling you I'd start by looking at the mission's JSON and maybe manually edit it to your requirements.

**Comment by Fogest** [comment]
My main issue is ones like "Reported Fraud" that can end up getting
dispatched only half of the time. So I know the missions json is correct as
it does work sometimes. It can go all night without dispatching it. I'll go
check and I'll have multiple units available to respond. And then in the
bots logs I believe it just doesn't pick up that the mission even is seen

**Comment by jackbayliss** [comment]
If it's not picking up the mission, you may have to re-generate the new missions. Let me know if it helps etc. As I've not had any other reports.

**Comment by jackbayliss** [comment]
Closing this as it seems the fix was to update the missions.json, feel free to reopen if it's not fixed 👍🏻

**Comment by Fogest** [comment]
The regenerating may have helped a bit. Editing the various json files also may have helped a bit. What also has been helping is that I have progressed more in the game and have more unit types available. As such the problem seems to be less of an issue now.

**Comment by jackbayliss** [comment]
@Fogest Good to know! I guess if you don't have a particular vehicle you could always add it into the vehicleignorelist.json - I'm glad its fixed for you anyway 👍🏻

**Comment by Fogest** [comment]
On the old version I was able to fix the problem partially also by having it share those missions with the alliance that way it would still dispatch them. A problem I noticed though is that it would keep re-dispatching more units to the call despite all those types of vehicles requirements already being met. I will have to see if this is a problem in the new version or not. I had disabled alliance sharing because of this.
